lib/cosmos/interfaces/protocols/terminated_stream_protocol.rb

```diff
@@ -28,58 +27,56 @@ class TerminatedStreamProtocol < StreamProtocol
     #   Packet.
     # @param discard_leading_bytes (see StreamProtocol#initialize)
     # @param sync_pattern (see StreamProtocol#initialize)
-    # @param fill_sync_pattern (see StreamProtocol#initialize)
-    def initialize(write_termination_characters,
-                   read_termination_characters,
-                   strip_read_termination = true,
-                   discard_leading_bytes = 0,
-                   sync_pattern = nil,
-                   fill_sync_pattern = false)
+    # @param fill_fields (see StreamProtocol#initialize)
+    def initialize(
+      write_termination_characters,
+      read_termination_characters,
+      strip_read_termination = true,
+      discard_leading_bytes = 0,
+      sync_pattern = nil,
+      fill_fields = false
+    )
       @write_termination_characters = write_termination_characters.hex_to_byte_string
       @read_termination_characters = read_termination_characters.hex_to_byte_string
       @strip_read_termination = ConfigParser.handle_true_false(strip_read_termination)
 
-      super(discard_leading_bytes, sync_pattern, fill_sync_pattern)
+      super(discard_leading_bytes, sync_pattern, fill_fields)
     end
 
-    # See StreamProtocol#pre_write_packet
-    def pre_write_packet(packet)
-      data = super(packet)
+    def write_data(data)
       raise "Packet contains termination characters!" if data.index(@write_termination_characters)
-
-      data = data.clone # Don't want to modify the actual packet buffer with the termination characters
+      data = super(data)
       @write_termination_characters.each_byte do |byte|
         data << byte
       end
-      data
+      return data
     end
 
     protected
 
     def reduce_to_single_packet
-      while true
-        index = @data.index(@read_termination_characters)
+      index = @data.index(@read_termination_characters)
 
-        # Reduce to packet data and setup current_data for next packet
-        if index
-          if index > 0
-            if @strip_read_termination
-              packet_data = @data[0..(index - 1)]
-            else
-              packet_data = @data[0..(index + @read_termination_characters.length - 1)]
-            end
+      # Reduce to packet data and setup current_data for next packet
+      if index
+        if index > 0
+          if @strip_read_termination
+            packet_data = @data[0..(index - 1)]
           else
+            packet_data = @data[0..(index + @read_termination_characters.length - 1)]
+          end
+        else # @data begins with the termination characters
+          if @strip_read_termination
             packet_data = ''
+          else # Keep everything
+            packet_data = @data[0..-1]
```

Comment:
No.  It says the data begins with the termination characters, not that @data ONLY contains the termination characters.  This should be packet_data = @data[0..(@read_termination_characters.length - 1)]